### PR TITLE
[make:subscriber] Improve MakeSubscriber to use KernelEvents constant instead hardcoded event

### PR DESCRIPTION
**Pull request by bdaler**
Use KernelEvents constant instead hardcoded event type in MakeSubscriber.
#744

### CODE REVIEW

File: src/Maker/MakeSubscriber.php [hunk 1]
```diff
@@ -115,4 +120,43 @@ public function generate(InputInterface $input, ConsoleStyle $io, Generator $gen
...
+     */
+    private function getEventConstant($event): string
+    {
+        $constants = $this->getKernelEventsConstants();
+
+        foreach ($constants as $name => $value) {
+            if ($value === $event) {
+                return 'KernelEvents::'.$name;
+            }
+        }
```
Comment (weaverryan) [inline src/Maker/MakeSubscriber.php:135]:
```php
$key = array_search($event, $constants);
if (false !== $key) 
   return sprintf('KernelEvents::%s', $name);
}

// ...
```

File: src/Maker/MakeSubscriber.php [hunk 2]
```diff
@@ -115,4 +120,43 @@ public function generate(InputInterface $input, ConsoleStyle $io, Generator $gen
     public function configureDependencies(DependencyBuilder $dependencies): void
     {
     }
+
+    /**
+     * @param $event
+     */
+    private function getEventConstant($event): string
```
Comment (weaverryan) [inline src/Maker/MakeSubscriber.php:127]:
```suggestion
    private function getEventConstant(string $event): string
```

And remove `@param`.

**Review by weaverryan — COMMENTED** [review]
I really like the approach - minor comments, and we need a test!

Also, in the interactive list that we show the users (when asking for which event they want), do we show the string name `kernel.controller` or the class name - `ControllerEvent`? We should probably show the class name.

**Comment by jrushlow** [comment]
>Also, in the interactive list that we show the users (when asking for which event they want), do we show the string name kernel.controller or the class name - ControllerEvent? We should probably show the class name.

This is what we have - 
image

Seems like the best of both worlds

File: src/Maker/MakeSubscriber.php [hunk 3]
```diff
@@ -87,6 +88,14 @@ public function generate(InputInterface $input, ConsoleStyle $io, Generator $gen
             EventSubscriberInterface::class,
         ]);
 
+        // Determine if we use a KernelEvents::CONSTANT or custom even name
+        if (null !== ($eventConstant = $this->getEventConstant($event))) {
+            $useStatements->addUseStatement(KernelEvents::class);
+            $eventName = $eventConstant;
+        } else {
+            $eventName = class_exists($event) ? sprintf('%s::class', $eventClassName) : sprintf('\'%s\'', $event);
+        }
```
[the remark this answers is not in the document]
Comment (jrushlow) [reply src/Maker/MakeSubscriber.php:97]: I don't think we can without adding more noise.. we have to handle a custom event like `\Symfony\MakerBundle\Generator::class` and assert that we subscribe to `Generator::class => onGenerator`. Side note - just added a testcase for this situation

File: src/Maker/MakeSubscriber.php [hunk 4]
```diff
@@ -115,4 +124,20 @@ public function generate(InputInterface $input, ConsoleStyle $io, Generator $gen
...
+            return sprintf('KernelEvents::%s', $name);
+        }
+
+        return null;
+    }
+
+    private function getKernelEventsConstants(): array
+    {
+        return (new \ReflectionClass(KernelEvents::class))->getConstants();
+    }
```
[the remark this answers is not in the document]
Comment (jrushlow) [reply src/Maker/MakeSubscriber.php]: done

**Pull request merged**